lib/spack/docs/contribution_guide.rst

```diff
@@ -1,247 +1,514 @@
+^^^^^^^^^^^^^^
+
+What if you made some changes to your local modified develop branch and already
+committed them, but later decided to contribute them to Spack? You can use
+cherry-picking to create a new branch with only these commits.
+
+First, check out your local modified develop branch:
 
 .. code-block:: console
 
-   $ git status
-   $ git diff
+   $ git checkout <your_modified_develop_branch>
 
-One can also run GUI to visualize the current changes
+Now, get the hashes of the commits you want from the output of:
 
 .. code-block:: console
 
-   $ git difftool
+   $ git log
 
-Next step is to rewrite the history by adding files and creating commits
+Next, create a new branch off of upstream ``develop`` and copy the commits
+that you want in your PR:
 
 .. code-block:: console
 
-   $ git add <files_to_be_part_of_commit>
-   $ git commit -m <descriptive_message>
+   $ git checkout develop
+   $ git pull upstream develop
+   $ git branch <descriptive_branch_name>
+   $ git checkout <descriptive_branch_name>
+   $ git cherry-pick <hash>
+   $ git push origin <descriptive_branch_name> --set-upstream
 
+Now you can create a PR from the web-interface of GitHub. The net result is as
+follows:
 
-Shall you need to split changes within a file into separate commits, use
+#. You patched your local version of Spack and can use it further.
+#. You "cherry-picked" these changes in a stand-alone branch and submitted it
+   as a PR upstream.
+
+Should you have several commits to contribute, you could follow the same
+procedure by getting hashes of all of them and cherry-picking to the PR branch.
+
+.. note::
+
+   It is important that whenever you change something that might be of
+   importance upstream, create a pull request as soon as possible. Do not wait
+   for weeks/months to do this, because:
+
+   #. you might forget why you modified certain files
+   #. it could get difficult to isolate this change into a stand-alone clean PR.
+
+^^^^^^^^
+Rebasing
+^^^^^^^^
+
+Other developers are constantly making contributions to Spack, possibly on the
+same files that your PR changed. If their PR is merged before yours, it can
+create a merge conflict. This means that your PR can no longer be automatically
+merged without a chance of breaking your changes. In this case, you will be
+asked to rebase on top of the latest upstream ``develop``.
+
+First, make sure your develop branch is up-to-date:
 
 .. code-block:: console
 
-   $ git add <file> -p
+   $ git checkout develop
+   $ git pull upstream develop
 
-After all changed files are committed, you can push the branch to your fork
-and create a PR
+Now, we need to switch to the branch you submitted for your PR and rebase it
+on top of develop:
+
+.. code-block:: console
+
+   $ git checkout <descriptive_branch_name>
+   $ git rebase develop
+
+This will likely crash when it comes to the conflicting commit. Edit the file
+that it says can't be merged automatically and resolve the conflict. Then, run:
```

Comment:
i find this very helpful once you are used to specific GUI of the tool. 

You can set mergetool by `git config --global merge.tool kdiff3` or use it only for a single merge `git mergetool --tool=kdiff3`. To check the list of tools `git mergetool --tool-help`. Here is some GUI docu for kdiff3 http://kdiff3.sourceforge.net/doc/merging.html 

Feel free to use some of none of this at all. :smile: